javascript/ql/src/semmle/javascript/Regexp.qll

```diff
@@ -404,3 +407,87 @@ module RegExpPatterns {
     result = "(?:com|org|edu|gov|uk|net|io)(?![a-z0-9])"
   }
 }
+
+/**
+ * Gets a node whose value may flow (inter-procedurally) to a position where it is interpreted
+ * as a part of a regular expression.
+ */
+private DataFlow::Node regExpSource(DataFlow::Node re, DataFlow::TypeBackTracker t) {
+  t.start() and
+  re = result and
+  isInterpretedAsRegExp(result)
+  or
+  exists(DataFlow::TypeBackTracker t2, DataFlow::Node succ | succ = regExpSource(re, t2) |
+    t2 = t.smallstep(result, succ)
+    or
+    any(TaintTracking::AdditionalTaintStep dts).step(result, succ) and
+    t = t2
+  )
+}
+
+/**
+ * Gets a node whose value may flow (inter-procedurally) to a position where it is interpreted
+ * as a part of a regular expression.
+ */
+private DataFlow::Node regExpSource(DataFlow::Node re) {
+  result = regExpSource(re, DataFlow::TypeBackTracker::end())
+}
+
+/**
+ * A node whose value may flow to a position where it is interpreted
+ * as a part of a regular expression.
+ */
+abstract class RegExpPatternSource extends DataFlow::Node {
+  /**
+   * Gets the pattern of this node.
+   */
+  abstract string getPattern();
+
+  /**
+   * Gets a regular expression object that is built from the value of this node.
+   */
+  abstract DataFlow::SourceNode getARegExpObject();
+}
+
+/**
+ * A regular expression literal, viewed as the pattern source for itself.
+ */
+class RegExpLiteralPatternSource extends RegExpPatternSource {
+  string pattern;
+
+  RegExpLiteralPatternSource() {
+    exists(string raw | raw = asExpr().(RegExpLiteral).getRoot().getRawValue() |
+      // hide the fact that `/` is escaped in the literal
+      pattern = raw.regexpReplaceAll("\\\\/", "/")
+    )
+  }
+
+  override string getPattern() { result = pattern }
+
+  override DataFlow::SourceNode getARegExpObject() { result = this }
+}
+
+/**
+ * A node whose string value may flow to a position where it is interpreted
+ * as a part of a regular expression.
+ */
+class StringRegExpPatternSource extends RegExpPatternSource {
```

Comment:
Could/should this class be `private`?

Reply:
It can't be private due to the use of class checks and `getAUse` in two places for whitelisting and alert presentation:

`js/regex/missing-regexp-anchor`:
```
          src.getARegExpObject().flowsTo(arg) or
          src.(StringRegExpPatternSource).getAUse() = arg
```
`js/incomplete-hostname-regexp`: 
```
    if re instanceof StringRegExpPatternSource
    then (
      kind = "string, which is used as a regular expression $@," and
      aux = re.(StringRegExpPatternSource).getAUse()
    ) else (
      kind = "regular expression" and aux = re
    )
```